nomos-da/network/core/src/protocols/replication/behaviour_v2.rs

```diff
@@ -0,0 +1,677 @@
+    /// Appends a message to the peer's pending queue. Lazily creates the queue
+    /// for the peer upon the first call.
+    fn enqueue_message(&mut self, peer_id: PeerId, message: DaMessage) {
+        self.messages.entry(peer_id).or_default().push_back(message);
+    }
+
+    /// ### Panics
+    ///
+    /// [`Self::enqueue_message`] must be called at least once for the peer
+    /// before calling this method or it will panic.
+    fn dequeue_message(&mut self, peer_id: &PeerId) -> Option<DaMessage> {
+        self.messages
+            .get_mut(peer_id)
+            .expect("Peer is in the map")
+            .pop_front()
+    }
+
+    /// Do internal housekeeping when a peer disconnects.
+    fn on_disconnected(&mut self, peer_id: &PeerId) {
+        self.last_scheduled = on_disconnected(&mut self.messages, self.last_scheduled, peer_id);
+    }
+
+    /// Find the next peer whose pending message can be scheduled for writing to
+    /// the peer's stream. Peers are iterated in a round-robin fashion.
+    fn next_peer(&mut self, is_stream_idle_fn: impl FnMut(&PeerId) -> bool) -> Option<PeerId> {
+        let next = next_peer(
+            self.last_scheduled.as_ref(),
+            &self.messages,
+            is_stream_idle_fn,
+        );
+        self.last_scheduled = next;
+        next
+    }
+}
+
+/// Nomos DA broadcast network behaviour.
+///
+/// This item handles the logic of the nomos da subnetworks broadcasting.
+/// DA subnetworks are a logical distribution of subsets.
+/// A node just connects and accepts connections to other nodes that are in the
+/// same subsets. A node forwards messages to all connected peers which are
+/// member of the addressed `SubnetworkId`.
+pub struct ReplicationBehaviour<Membership> {
+    /// Local peer Id, related to the libp2p public key
+    local_peer_id: PeerId,
+    /// Underlying stream behaviour
+    stream_behaviour: libp2p_stream::Behaviour,
+    /// Used to open new outbound streams from the stream behaviour
+    control: Control,
+    /// Provides inbound streams that are accepted by the stream behaviour
+    incoming_streams: IncomingStreams,
+    /// Holds tasks for reading messages from incoming streams
+    incoming_tasks: FuturesUnordered<IncomingTask>,
+    /// Membership handler, membership handles the subsets logics on who is
+    /// where in the nomos DA subnetworks
+    membership: Membership,
+    /// Currently connected peers
+    connected: HashSet<PeerId>,
+    /// Pending outbound messages are stored here, they are then consumed by
+    /// each respective outbound stream
+    pending_outbound: PendingOutbound,
+    /// Indicates which outbound streams are currently idle or busy
+    outbound_streams: HashMap<PeerId, StreamState>,
+    /// Holds tasks for writing messages to the outbound streams
+    outbound_tasks: FuturesUnordered<OutboundTask>,
+    /// Seen messages cache holds a record of seen messages
+    seen_message_cache: IndexSet<(Vec<u8>, SubnetworkId)>,
```

Comment:
This probably should be a proper chache. Otherwise it will explode with time. Some RLU with a proper policy. Can be done in a different PR but lets open an issue to track it.